File: Linux/CSharp/Storms.cs
using System;
using System.CommandLine;
using System.Threading.Tasks;
using System.Linq;
using StormSearch.Utils;
using StormSearch.Search.Minecraft;
using StormSearch.Search.Steam;
using StormSearch.Search.XBOX;

namespace StormSearch
{
    public class Storms
    {
        private static readonly string[] AvailablePlatforms = { "minecraft", "steam", "xbox", "all" };
        
        public static async Task<int> RunStormsCommand(string[] args)
        {
            // парсинг аргументов в стиле sherlock
            if (args.Length == 0)
            {
                ShowStormsHelp();
                return 1;
            }

            var username = args[0];
            var platforms = new List<string>();
            var verbose = false;
            var output = "";

            // парсинг флагов
            for (int i = 1; i < args.Length; i++)
            {
                var arg = args[i].ToLower();
                
                switch (arg)
                {
                    case "-v":
                    case "--verbose":
                        verbose = true;
                        break;
                    case "-o":
                    case "--output":
                        if (i + 1 < args.Length)
                        {
                            output = args[i + 1];
                            i++; // пропустить следующий аргумент
                        }
                        break;
                    case "-p":
                    case "--platform":
                        if (i + 1 < args.Length)
                        {
                            var platformArg = args[i + 1].ToLower();
                            if (platformArg == "all")
                            {
                                platforms.AddRange(AvailablePlatforms.Where(p => p != "all"));
                            }
                            else if (AvailablePlatforms.Contains(platformArg))
                            {
                                platforms.Add(platformArg);
                            }
                            else
                            {
                                Colors.PrintError($"неизвестная платформа: {platformArg}");
                                return 1;
                            }
                            i++; // пропустить следующий аргумент
                        }
                        break;
                    case "-h":
                    case "--help":
                        ShowStormsHelp();
                        return 0;
                    default:
                        if (arg.StartsWith("-"))
                        {
                            Colors.PrintError($"неизвестный флаг: {arg}");
                            return 1;
                        }
                        break;
                }
            }

            // если платформы не указаны, ищем по всем
            if (platforms.Count == 0)
            {
                platforms.AddRange(AvailablePlatforms.Where(p => p != "all"));
            }

            // вывод информации о поиске
            Console.WriteLine($"[*] storm search - поиск пользователя: {username}");
            if (verbose)
            {
                Console.WriteLine($"[*] платформы: {string.Join(", ", platforms)}");
                if (!string.IsNullOrEmpty(output))
                {
                    Console.WriteLine($"[*] вывод в файл: {output}");
                }
            }
            Console.WriteLine("═".PadRight(60, '═'));

            var results = new Dictionary<string, object>();

            // поиск по указанным платформам
            foreach (var platform in platforms)
            {
                try
                {
                    if (verbose)
                    {
                        Console.WriteLine($"[*] поиск в {platform}...");
                    }

                    object result = null;
                    
                    switch (platform)
                    {
                        case "minecraft":
                            result = await SearchMinecraftUser(username, verbose);
                            break;
                        case "steam":
                            result = await SearchSteamUser(username, verbose);
                            break;
                        case "xbox":
                            result = await SearchXboxUser(username, verbose);
                            break;
                    }

                    if (result != null)
                    {
                        results[platform] = result;
                    }

                    Console.WriteLine();
                    await Task.Delay(1000); // задержка между запросами
                }
                catch (Exception ex)
                {
                    Colors.PrintError($"ошибка при поиске в {platform}: {ex.Message}");
                }
            }

            // вывод результатов
            PrintStormsResults(username, results, verbose);

            // сохранение в файл если указано
            if (!string.IsNullOrEmpty(output))
            {
                await SaveResultsToFile(results, output, verbose);
            }

            return 0;
        }

        private static async Task<object> SearchMinecraftUser(string username, bool verbose)
        {
            try
            {
                var searcher = new MinecraftSearcher();
                var result = await searcher.SearchPlayerAsync(username);
                
                if (verbose && result != null)
                {
                    Console.WriteLine($"[+] minecraft: найден - {result.Username}");
                }
                
                return result;
            }
            catch (Exception ex)
            {
                if (verbose)
                {
                    Colors.PrintError($"minecraft ошибка: {ex.Message}");
                }
                return null;
            }
        }

        private static async Task<object> SearchSteamUser(string username, bool verbose)
        {
            try
            {
                var searcher = new SteamSearcher();
                var result = await searcher.SearchPlayerAsync(username);
                
                if (verbose && result != null)
                {
                    Console.WriteLine($"[+] steam: найден - {result.Username}");
                }
                
                return result;
            }
            catch (Exception ex)
            {
                if (verbose)
                {
                    Colors.PrintError($"steam ошибка: {ex.Message}");
                }
                return null;
            }
        }

        private static async Task<object> SearchXboxUser(string username, bool verbose)
        {
            try
            {
                var searcher = new XboxSearcher();
                var result = await searcher.SearchPlayerAsync(username);
                
                if (verbose && result != null)
                {
                    Console.WriteLine($"[+] xbox: найден - {result.Username}");
                }
                
                return result;
            }
            catch (Exception ex)
            {
                if (verbose)
                {
                    Colors.PrintError($"xbox ошибка: {ex.Message}");
                }
                return null;
            }
        }

        private static void PrintStormsResults(string username, Dictionary<string, object> results, bool verbose)
        {
            Console.WriteLine("═".PadRight(60, '═'));
            Console.WriteLine($"[*] результаты поиска для: {username}");
            Console.WriteLine("═".PadRight(60, '═'));
            Console.WriteLine();

            if (results.Count == 0)
            {
                Colors.PrintError("пользователь не найден ни на одной платформе");
                return;
            }

            var foundCount = 0;
            foreach (var kvp in results)
            {
                var platform = kvp.Key;
                var result = kvp.Value;

                if (result is SearchResult searchResult && searchResult.Found)
                {
                    foundCount++;
                    Colors.PrintSuccess($"{platform}: найден");
                    
                    Console.WriteLine($"   пользователь: {searchResult.Username}");
                    if (!string.IsNullOrEmpty(searchResult.Id))
                    {
                        Console.WriteLine($"   id: {searchResult.Id}");
                    }
                    
                    if (searchResult.AdditionalInfo.Count > 0)
                    {
                        Console.WriteLine("   дополнительная информация:");
                        foreach (var info in searchResult.AdditionalInfo.Take(3)) // показываем первые 3 поля
                        {
                            Console.WriteLine($"     {info.Key}: {info.Value}");
                        }
                        if (searchResult.AdditionalInfo.Count > 3)
                        {
                            Console.WriteLine($"     ... и еще {searchResult.AdditionalInfo.Count - 3} полей");
                        }
                    }
                }
                else
                {
                    Colors.PrintWarning($"{platform}: не найден");
                }
                Console.WriteLine();
            }

            Console.WriteLine("═".PadRight(60, '═'));
            Console.WriteLine($"[*] статистика: найдено на {foundCount} из {results.Count} платформ");
            
            if (verbose)
            {
                Console.WriteLine($"[*] выполнено за: {DateTime.Now:HH:mm:ss}");
            }
        }

        private static async Task SaveResultsToFile(Dictionary<string, object> results, string filename, bool verbose)
        {
            try
            {
                var json = System.Text.Json.JsonSerializer.Serialize(results, new System.Text.Json.JsonSerializerOptions 
                { 
                    WriteIndented = true 
                });
                
                await File.WriteAllTextAsync(filename, json);
                
                if (verbose)
                {
                    Colors.PrintSuccess($"результаты сохранены в: {filename}");
                }
            }
            catch (Exception ex)
            {
                Colors.PrintError($"ошибка сохранения файла: {ex.Message}");
            }
        }

        private static void ShowStormsHelp()
        {
            Console.WriteLine();
            Colors.PrintInfo("storm search - утилита поиска никнеймов (стиль sherlock)");
            Console.WriteLine();
            Console.WriteLine("использование:");
            Console.WriteLine("  storms <username> [опции]");
            Console.WriteLine();
            Console.WriteLine("опции:");
            Console.WriteLine("  -p, --platform <платформа>  указать платформу (minecraft, steam, xbox, all)");
            Console.WriteLine("  -v, --verbose             подробный вывод");
            Console.WriteLine("  -o, --output <файл>      сохранить результаты в файл");
            Console.WriteLine("  -h, --help                 показать эту справку");
            Console.WriteLine();
            Console.WriteLine("платформы:");
            Console.WriteLine("  minecraft    - поиск игроков minecraft");
            Console.WriteLine("  steam        - поиск профилей steam");
            Console.WriteLine("  xbox         - поиск профилей xbox");
            Console.WriteLine("  all          - поиск по всем платформам (по умолчанию)");
            Console.WriteLine();
            Console.WriteLine("примеры:");
            Console.WriteLine("  storms abrakadam");
            Console.WriteLine("  storms player123 -p minecraft");
            Console.WriteLine("  storms username -v -o results.json");
            Console.WriteLine("  storms gamer -p steam,minecraft -v");
            Console.WriteLine();
            Console.WriteLine("примечание: без указания платформ поиск выполняется по всем доступным платформам");
        }
    }
}

File: Linux/JS/storms.js
#!/usr/bin/env node

// storms команда в стиле sherlock для быстрого поиска
const minecraft_scraper = require('./scrapers/minecraft-scraper');
const steam_scraper = require('./scrapers/steam-scraper');
const xbox_scraper = require('./scrapers/xbox-scraper');
const fs = require('fs').promises;

class storms_command {
    constructor() {
        this.scrapers = {
            'minecraft': new minecraft_scraper(),
            'steam': new steam_scraper(),
            'xbox': new xbox_scraper()
        };
        this.available_platforms = ['minecraft', 'steam', 'xbox', 'all'];
    }

    // основная команда storms
    async run(args) {
        try {
            // парсинг аргументов
            const parsed = this.parse_args(args);
            
            if (parsed.help) {
                this.show_help();
                return 0;
            }

            if (!parsed.username) {
                console.log('[-] ошибка: необходимо указать имя пользователя');
                console.log('[*] использование: storms <username> [опции]');
                console.log('[*] для помощи: storms --help');
                return 1;
            }

            // вывод информации о поиске
            console.log(`[*] storm search - поиск пользователя: ${parsed.username}`);
            if (parsed.verbose) {
                console.log(`[*] платформы: ${parsed.platforms.join(', ')}`);
                if (parsed.output) {
                    console.log(`[*] вывод в файл: ${parsed.output}`);
                }
            }
            console.log('═'.repeat(60));

            const results = {};

            // поиск по указанным платформам
            for (const platform of parsed.platforms) {
                try {
                    if (parsed.verbose) {
                        console.log(`\n[*] поиск в ${platform}...`);
                    }

                    const result = await this.search_platform(parsed.username, platform);
                    if (result) {
                        results[platform] = result;
                    }

                    // задержка между запросами
                    await this.sleep(1000);
                } catch (error) {
                    console.log(`[-] ошибка при поиске в ${platform}: ${error.message}`);
                }
            }

            // вывод результатов
            this.print_results(parsed.username, results, parsed.verbose);

            // сохранение в файл если указано
            if (parsed.output) {
                await this.save_results(results, parsed.output, parsed.verbose);
            }

            return 0;
        } catch (error) {
            console.log(`[-] критическая ошибка: ${error.message}`);
            return 1;
        }
    }

    // парсинг аргументов командной строки
    parse_args(args) {
        const parsed = {
            username: null,
            platforms: ['minecraft', 'steam', 'xbox'], // по умолчанию все
            verbose: false,
            output: null,
            help: false
        };

        for (let i = 0; i < args.length; i++) {
            const arg = args[i].toLowerCase();
            
            switch (arg) {
                case '-v':
                case '--verbose':
                    parsed.verbose = true;
                    break;
                case '-o':
                case '--output':
                    if (i + 1 < args.length) {
                        parsed.output = args[i + 1];
                        i++; // пропустить следующий аргумент
                    }
                    break;
                case '-p':
                case '--platform':
                    if (i + 1 < args.length) {
                        const platform_arg = args[i + 1].toLowerCase();
                        if (platform_arg === 'all') {
                            parsed.platforms = ['minecraft', 'steam', 'xbox'];
                        } else if (this.available_platforms.includes(platform_arg)) {
                            parsed.platforms = [platform_arg];
                        } else {
                            console.log(`[-] ошибка: неизвестная платформа ${platform_arg}`);
                            process.exit(1);
                        }
                        i++; // пропустить следующий аргумент
                    }
                    break;
                case '-h':
                case '--help':
                    parsed.help = true;
                    break;
                default:
                    if (!arg.startsWith('-') && !parsed.username) {
                        // первый аргумент без дефиса - это имя пользователя
                        parsed.username = arg;
                    } else if (arg.startsWith('-')) {
                        console.log(`[-] ошибка: неизвестный флаг ${arg}`);
                        process.exit(1);
                    }
                    break;
            }
        }

        return parsed;
    }

    // поиск на конкретной платформе
    async search_platform(username, platform) {
        if (!this.scrapers[platform]) {
            console.log(`[-] ошибка: неизвестная платформа ${platform}`);
            return null;
        }

        try {
            let result;
            switch (platform) {
                case 'minecraft':
                    result = await this.scrapers[platform].search_player(username);
                    break;
                case 'steam':
                    result = await this.scrapers[platform].search_player(username);
                    break;
                case 'xbox':
                    result = await this.scrapers[platform].search_player(username);
                    break;
                default:
                    result = null;
            }

            return result;
        } catch (error) {
            console.log(`[-] ошибка при поиске в ${platform}: ${error.message}`);
            return null;
        }
    }

    // вывод результатов поиска
    print_results(username, results, verbose = false) {
        console.log('\n' + '═'.repeat(60));
        console.log(`[*] результаты поиска для: ${username}`);
        console.log('═'.repeat(60));
        console.log();

        if (Object.keys(results).length === 0) {
            console.log('[-] пользователь не найден ни на одной платформе');
            return;
        }

        const found_count = Object.values(results).filter(r => r.found).length;
        const total_count = Object.keys(results).length;

        for (const [platform, result] of Object.entries(results)) {
            const status = result.found ? '[+] найден' : '[-] не найден';
            console.log(`${platform}: ${status}`);
            
            if (result.found) {
                console.log(`   пользователь: ${result.username || result.gamertag}`);
                
                if (result.uuid) {
                    console.log(`   uuid: ${result.uuid}`);
                }
                
                if (result.steam_id) {
                    console.log(`   steam id: ${result.steam_id}`);
                }
                
                if (result.xuid) {
                    console.log(`   xuid: ${result.xuid}`);
                }

                // вывод дополнительной информации
                if (result.profile_info && Object.keys(result.profile_info).length > 0) {
                    console.log('   информация о профиле:');
                    for (const [key, value] of Object.entries(result.profile_info)) {
                        if (verbose || ['persona_name', 'gamertag', 'display_name'].includes(key)) {
                            console.log(`     ${key}: ${value}`);
                        }
                    }
                    if (!verbose && Object.keys(result.profile_info).length > 3) {
                        console.log(`     ... и еще ${Object.keys(result.profile_info).length - 3} полей`);
                    }
                }

                if (result.name_history) {
                    console.log(`   история имен: ${result.name_history.history_string}`);
                }

                if (result.skin_info) {
                    console.log('   информация о скине: доступна');
                }

                if (result.games_info && Object.keys(result.games_info).length > 0) {
                    console.log('   информация об играх: доступна');
                }
            }
            console.log();
        }

        console.log('═'.repeat(60));
        console.log(`[*] статистика: найдено на ${found_count} из ${total_count} платформ`);
        
        if (verbose) {
            console.log(`[*] выполнено за: ${new Date().toLocaleTimeString()}`);
        }
    }

    // сохранение результатов в файл
    async save_results(results, filename, verbose = false) {
        try {
            const json = JSON.stringify(results, null, 2);
            await fs.writeFile(filename, json);
            
            if (verbose) {
                console.log(`[+] результаты сохранены в ${filename}`);
            }
        } catch (error) {
            console.log(`[-] ошибка сохранения файла: ${error.message}`);
        }
    }

    // вывод справки
    show_help() {
        console.log();
        console.log('storm search - утилита поиска никнеймов (стиль sherlock)');
        console.log();
        console.log('использование:');
        console.log('  storms <username> [опции]');
        console.log();
        console.log('опции:');
        console.log('  -p, --platform <платформа>  указать платформу (minecraft, steam, xbox, all)');
        console.log('  -v, --verbose             подробный вывод');
        console.log('  -o, --output <файл>      сохранить результаты в файл');
        console.log('  -h, --help                 показать эту справку');
        console.log();
        console.log('платформы:');
        console.log('  minecraft    - поиск игроков minecraft');
        console.log('  steam        - поиск профилей steam');
        console.log('  xbox         - поиск профилей xbox');
        console.log('  all          - поиск по всем платформам (по умолчанию)');
        console.log();
        console.log('примеры:');
        console.log('  storms abrakadam');
        console.log('  storms player123 -p minecraft');
        console.log('  storms username -v -o results.json');
        console.log('  storms gamer -p steam,minecraft -v');
        console.log();
        console.log('примечание: без указания платформ поиск выполняется по всем доступным платформам');
        console.log();
    }

    // задержка
    sleep(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }
}

// обработка командной строки
async function main() {
    const args = process.argv.slice(2);
    const storms = new storms_command();
    
    const exit_code = await storms.run(args);
    process.exit(exit_code);
}

// запуск
if (require.main === module) {
    main().catch(console.error);
}

module.exports = storms_command;
